daemon/runtime_unix.go

```diff
@@ -134,3 +137,35 @@ func (daemon *Daemon) getRuntime(name string) (*types.Runtime, error) {
 
 	return rt, nil
 }
+
+// isValidC8dRuntimeName tests whether name
+// is a well-formed ContainerD runtime name,
+// with additional Docker-specific restrictions.
+//
+// A runtime name containing slash characters is interpreted by ContainerD
+// as the path to a runtime binary.
+// If we allowed this, anyone with Engine API access could get ContainerD
+// to execute an arbitrary binary as root.
+// Although Engine API access is already equivalent to root on the host,
+// the runtime name has not historically been a vector
+// to run arbitrary code as root
+// so users are not expecting it to become one.
+//
+// This restriction is not configurable.
+// There are viable workarounds for legitimate use cases:
+// administrators and runtime developers can make runtimes available
+// for use with Docker by installing them onto $PATH
+// following the [binary naming convention] for ContainerD Runtime v2.
+//
+// [binary naming convention]: https://github.com/containerd/containerd/blob/main/runtime/v2/README.md#binary-naming
+func isValidC8dRuntimeName(name string) bool {
+	// ContainerD uses a rather permissive test to validate runtime names.
+	// Any name which meets either requirement is valid:
+	//   - filepath.IsAbs(name), or
+	//   - the name contains at least one '.' character and no '/' characters.
+	// The name "." is technically valid per ContainerD's implementation:
+	// it would resolve to a binary named "containerd-shim---".
+	// https://github.com/containerd/containerd/blob/11ded166c15f92450958078cd13c6d87131ec563/runtime/v2/manager.go#L297-L317
+	// https://github.com/containerd/containerd/blob/11ded166c15f92450958078cd13c6d87131ec563/runtime/v2/shim/util.go#L83-L93
+	return !strings.ContainsRune(name, '/') && strings.ContainsRune(name, '.')
```

Comment:
`filepath.Separator`?

Reply:
[ContainerD does a substring check against a literal `/` character.](https://github.com/containerd/containerd/blob/11ded166c15f92450958078cd13c6d87131ec563/runtime/v2/manager.go#L307-L310) I could add a `!filepath.IsAbs(name)` clause to cover the absolute-path-on-Windows case.